# Assignment-06-Logistic Regression

# '''The steps involved in logistic regression are:
Define the problem: Identify the dependent variable and independent variables and determine if the problem is a binary classification problem.

Data preparation: Clean and preprocess the data, and make sure the data is suitable for logistic regression.

Fitting Logistic Regression to the Training set.

Predicting the test result.

Test accuracy of the result (Creation of Confusion matrix).

Visualizing the test set result (optional).

Logistic regression has two phases: training and test.'''

In [2]:
# Importing the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score

In [4]:
# Importing the dataset
bank_data=pd.read_csv(r'C:\Users\BIS\Desktop\DATA SCIENCE ASSIGNMENTS\Logistics Regression\bank-full.csv',sep=";")
bank_data

age           job   marital  education default  balance housing loan  \
0       58    management   married   tertiary      no     2143     yes   no   
1       44    technician    single  secondary      no       29     yes   no   
2       33  entrepreneur   married  secondary      no        2     yes  yes   
3       47   blue-collar   married    unknown      no     1506     yes   no   
4       33       unknown    single    unknown      no        1      no   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
45206   51    technician   married   tertiary      no      825      no   no   
45207   71       retired  divorced    primary      no     1729      no   no   
45208   72       retired   married  secondary      no     5715      no   no   
45209   57   blue-collar   married  secondary      no      668      no   no   
45210   37  entrepreneur   married  secondary      no     2971      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
0        unknown    5   may       261         1     -1         0  unknown   no  
1        unknown    5   may       151         1     -1         0  unknown   no  
2        unknown    5   may        76         1     -1         0  unknown   no  
3        unknown    5   may        92         1     -1         0  unknown   no  
4        unknown    5   may       198         1     -1         0  unknown   no  
...          ...  ...   ...       ...       ...    ...       ...      ...  ...  
45206   cellular   17   nov       977         3     -1         0  unknown  yes  
45207   cellular   17   nov       456         2     -1         0  unknown  yes  
45208   cellular   17   nov      1127         5    184         3  success  yes  
45209  telephone   17   nov       508         4     -1         0  unknown   no  
45210   cellular   17   nov       361         2    188        11    other   no  

[45211 rows x 17 columns]

## EDA

In [5]:
bank_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    45211 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null  int64 
 14  previous   45211 non-null  int64 
 15  poutcome   45211 non-null  object
 16  y          45211 non-null  object
dtypes: int64(7), object(10)
memory usage: 5.9+ MB


# Data preparation

In [6]:
# One-Hot Encoding of categrical variables
bank_data1=pd.get_dummies(bank_data,columns=["job","marital","education","contact","month","poutcome"])
bank_data1

age default  balance housing loan  day  duration  campaign  pdays  \
0       58      no     2143     yes   no    5       261         1     -1   
1       44      no       29     yes   no    5       151         1     -1   
2       33      no        2     yes  yes    5        76         1     -1   
3       47      no     1506     yes   no    5        92         1     -1   
4       33      no        1      no   no    5       198         1     -1   
...    ...     ...      ...     ...  ...  ...       ...       ...    ...   
45206   51      no      825      no   no   17       977         3     -1   
45207   71      no     1729      no   no   17       456         2     -1   
45208   72      no     5715      no   no   17      1127         5    184   
45209   57      no      668      no   no   17       508         4     -1   
45210   37      no     2971      no   no   17       361         2    188   

       previous  ... month_jun  month_mar  month_may  month_nov  month_oct  \
0             0  ...         0          0          1          0          0   
1             0  ...         0          0          1          0          0   
2             0  ...         0          0          1          0          0   
3             0  ...         0          0          1          0          0   
4             0  ...         0          0          1          0          0   
...         ...  ...       ...        ...        ...        ...        ...   
45206         0  ...         0          0          0          1          0   
45207         0  ...         0          0          0          1          0   
45208         3  ...         0          0          0          1          0   
45209         0  ...         0          0          0          1          0   
45210        11  ...         0          0          0          1          0   

       month_sep  poutcome_failure  poutcome_other  poutcome_success  \
0              0                 0               0                 0   
1              0                 0               0                 0   
2              0                 0               0                 0   
3              0                 0               0                 0   
4              0                 0               0                 0   
...          ...               ...             ...               ...   
45206          0                 0               0                 0   
45207          0                 0               0                 0   
45208          0                 0               0                 1   
45209          0                 0               0                 0   
45210          0                 0               1                 0   

       poutcome_unknown  
0                     1  
1                     1  
2                     1  
3                     1  
4                     1  
...                 ...  
45206                 1  
45207                 1  
45208                 0  
45209                 1  
45210                 0  

[45211 rows x 49 columns]

In [7]:
# To see all columns
pd.set_option("display.max.columns", None)
bank_data1

age default  balance housing loan  day  duration  campaign  pdays  \
0       58      no     2143     yes   no    5       261         1     -1   
1       44      no       29     yes   no    5       151         1     -1   
2       33      no        2     yes  yes    5        76         1     -1   
3       47      no     1506     yes   no    5        92         1     -1   
4       33      no        1      no   no    5       198         1     -1   
...    ...     ...      ...     ...  ...  ...       ...       ...    ...   
45206   51      no      825      no   no   17       977         3     -1   
45207   71      no     1729      no   no   17       456         2     -1   
45208   72      no     5715      no   no   17      1127         5    184   
45209   57      no      668      no   no   17       508         4     -1   
45210   37      no     2971      no   no   17       361         2    188   

       previous    y  job_admin.  job_blue-collar  job_entrepreneur  \
0             0   no           0                0                 0   
1             0   no           0                0                 0   
2             0   no           0                0                 1   
3             0   no           0                1                 0   
4             0   no           0                0                 0   
...         ...  ...         ...              ...               ...   
45206         0  yes           0                0                 0   
45207         0  yes           0                0                 0   
45208         3  yes           0                0                 0   
45209         0   no           0                1                 0   
45210        11   no           0                0                 1   

       job_housemaid  job_management  job_retired  job_self-employed  \
0                  0               1            0                  0   
1                  0               0            0                  0   
2                  0               0            0                  0   
3                  0               0            0                  0   
4                  0               0            0                  0   
...              ...             ...          ...                ...   
45206              0               0            0                  0   
45207              0               0            1                  0   
45208              0               0            1                  0   
45209              0               0            0                  0   
45210              0               0            0                  0   

       job_services  job_student  job_technician  job_unemployed  job_unknown  \
0                 0            0               0               0            0   
1                 0            0               1               0            0   
2                 0            0               0               0            0   
3                 0            0               0               0            0   
4                 0            0               0               0            1   
...             ...          ...             ...             ...          ...   
45206             0            0               1               0            0   
45207             0            0               0               0            0   
45208             0            0               0               0            0   
45209             0            0               0               0            0   
45210             0            0               0               0            0   

       marital_divorced  marital_married  marital_single  education_primary  \
0                     0                1               0                  0   
1                     0                0               1                  0   
2                     0                1               0                  0   
3                     0                1               0                  0   
4       

In [8]:
bank_data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 49 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   age                  45211 non-null  int64 
 1   default              45211 non-null  object
 2   balance              45211 non-null  int64 
 3   housing              45211 non-null  object
 4   loan                 45211 non-null  object
 5   day                  45211 non-null  int64 
 6   duration             45211 non-null  int64 
 7   campaign             45211 non-null  int64 
 8   pdays                45211 non-null  int64 
 9   previous             45211 non-null  int64 
 10  y                    45211 non-null  object
 11  job_admin.           45211 non-null  uint8 
 12  job_blue-collar      45211 non-null  uint8 
 13  job_entrepreneur     45211 non-null  uint8 
 14  job_housemaid        45211 non-null  uint8 
 15  job_management       45211 non-null  uint8 
 16  job_

In [9]:
# Custom Binary Encoding of Binary o/p variables 
bank_data1['default'] = np.where(bank_data1['default'].str.contains("yes"), 1, 0)
bank_data1['housing'] = np.where(bank_data1['housing'].str.contains("yes"), 1, 0)
bank_data1['loan'] = np.where(bank_data1['loan'].str.contains("yes"), 1, 0)
bank_data1['y'] = np.where(bank_data1['y'].str.contains("yes"), 1, 0)
bank_data1

age  default  balance  housing  loan  day  duration  campaign  pdays  \
0       58        0     2143        1     0    5       261         1     -1   
1       44        0       29        1     0    5       151         1     -1   
2       33        0        2        1     1    5        76         1     -1   
3       47        0     1506        1     0    5        92         1     -1   
4       33        0        1        0     0    5       198         1     -1   
...    ...      ...      ...      ...   ...  ...       ...       ...    ...   
45206   51        0      825        0     0   17       977         3     -1   
45207   71        0     1729        0     0   17       456         2     -1   
45208   72        0     5715        0     0   17      1127         5    184   
45209   57        0      668        0     0   17       508         4     -1   
45210   37        0     2971        0     0   17       361         2    188   

       previous  y  job_admin.  job_blue-collar  job_entrepreneur  \
0             0  0           0                0                 0   
1             0  0           0                0                 0   
2             0  0           0                0                 1   
3             0  0           0                1                 0   
4             0  0           0                0                 0   
...         ... ..         ...              ...               ...   
45206         0  1           0                0                 0   
45207         0  1           0                0                 0   
45208         3  1           0                0                 0   
45209         0  0           0                1                 0   
45210        11  0           0                0                 1   

       job_housemaid  job_management  job_retired  job_self-employed  \
0                  0               1            0                  0   
1                  0               0            0                  0   
2                  0               0            0                  0   
3                  0               0            0                  0   
4                  0               0            0                  0   
...              ...             ...          ...                ...   
45206              0               0            0                  0   
45207              0               0            1                  0   
45208              0               0            1                  0   
45209              0               0            0                  0   
45210              0               0            0                  0   

       job_services  job_student  job_technician  job_unemployed  job_unknown  \
0                 0            0               0               0            0   
1                 0            0               1               0            0   
2                 0            0               0               0            0   
3                 0            0               0               0            0   
4                 0            0               0               0            1   
...             ...          ...             ...             ...          ...   
45206             0            0               1               0            0   
45207             0            0               0               0            0   
45208             0            0               0               0            0   
45209             0            0               0               0            0   
45210             0            0               0               0            0   

       marital_divorced  marital_married  marital_single  education_primary  \
0                     0                1               0                  0   
1                     0                0               1                  0   
2                     0                1               0                  0   
3                     0                1               0                  0

In [10]:
bank_data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 49 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   age                  45211 non-null  int64
 1   default              45211 non-null  int32
 2   balance              45211 non-null  int64
 3   housing              45211 non-null  int32
 4   loan                 45211 non-null  int32
 5   day                  45211 non-null  int64
 6   duration             45211 non-null  int64
 7   campaign             45211 non-null  int64
 8   pdays                45211 non-null  int64
 9   previous             45211 non-null  int64
 10  y                    45211 non-null  int32
 11  job_admin.           45211 non-null  uint8
 12  job_blue-collar      45211 non-null  uint8
 13  job_entrepreneur     45211 non-null  uint8
 14  job_housemaid        45211 non-null  uint8
 15  job_management       45211 non-null  uint8
 16  job_retired          4

## Model Building

In [11]:
# Dividing our data into input and output variables
X = bank_data1.drop(columns='y')
X

age  default  balance  housing  loan  day  duration  campaign  pdays  \
0       58        0     2143        1     0    5       261         1     -1   
1       44        0       29        1     0    5       151         1     -1   
2       33        0        2        1     1    5        76         1     -1   
3       47        0     1506        1     0    5        92         1     -1   
4       33        0        1        0     0    5       198         1     -1   
...    ...      ...      ...      ...   ...  ...       ...       ...    ...   
45206   51        0      825        0     0   17       977         3     -1   
45207   71        0     1729        0     0   17       456         2     -1   
45208   72        0     5715        0     0   17      1127         5    184   
45209   57        0      668        0     0   17       508         4     -1   
45210   37        0     2971        0     0   17       361         2    188   

       previous  job_admin.  job_blue-collar  job_entrepreneur  job_housemaid  \
0             0           0                0                 0              0   
1             0           0                0                 0              0   
2             0           0                0                 1              0   
3             0           0                1                 0              0   
4             0           0                0                 0              0   
...         ...         ...              ...               ...            ...   
45206         0           0                0                 0              0   
45207         0           0                0                 0              0   
45208         3           0                0                 0              0   
45209         0           0                1                 0              0   
45210        11           0                0                 1              0   

       job_management  job_retired  job_self-employed  job_services  \
0                   1            0                  0             0   
1                   0            0                  0             0   
2                   0            0                  0             0   
3                   0            0                  0             0   
4                   0            0                  0             0   
...               ...          ...                ...           ...   
45206               0            0                  0             0   
45207               0            1                  0             0   
45208               0            1                  0             0   
45209               0            0                  0             0   
45210               0            0                  0             0   

       job_student  job_technician  job_unemployed  job_unknown  \
0                0               0               0            0   
1                0               1               0            0   
2                0               0               0            0   
3                0               0               0            0   
4                0               0               0            1   
...            ...             ...             ...          ...   
45206            0               1               0            0   
45207            0               0               0            0   
45208            0               0               0            0   
45209            0               0               0            0   
45210            0               0               0            0   

       marital_divorced  marital_married  marital_single  education_primary  \
0                     0                1               0                  0   
1                     0                0               1                  0   
2                     0                1               0                  0   
3                     0                1               0                  0   
4                     0         

In [12]:
y_data = bank_data1.loc[:,'y']
y_data

0        0
1        0
2        0
3        0
4        0
        ..
45206    1
45207    1
45208    1
45209    0
45210    0
Name: y, Length: 45211, dtype: int32

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X,y_data,test_size=0.2)
X_train, X_test, y_train, y_test 

(       age  default  balance  housing  loan  day  duration  campaign  pdays  \
 20467   36        0     1659        0     0   12       126         2     -1   
 3123    38        0      360        1     0   14       220         2     -1   
 10899   37        0      942        0     0   17        79         1     -1   
 29459   58        0     3401        1     0    3        84         3    265   
 33154   52        0     1430        1     0   20       121         4     -1   
 ...    ...      ...      ...      ...   ...  ...       ...       ...    ...   
 36952   40        0      949        1     0   12       285         3    287   
 32059   40        0     1175        0     0   14       184         1     -1   
 28574   45        0      -55        0     0   29        78         3    197   
 28503   26        0     1274        0     0   29       475         2     -1   
 12546   52        0       51        0     0    3        88         1     -1   
 
        previous  job_admin.  job_blue

In [15]:
# Logistic regression model
classifier=LogisticRegression()
classifier.fit(X_train,y_train)

C:\Users\BIS\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

## Model Predictions

In [16]:
# Predict for x dataset
y_pred=classifier.predict(X_train)
y_pred

array([0, 0, 0, ..., 0, 0, 0])

In [19]:
y_pred_test = classifier.predict(X_test)
y_pred_test

array([0, 0, 0, ..., 0, 0, 0])

# Confusion_Matrix

In [26]:
# Confusion Matrix for the model accuracy
confusion_matrix(y_train,y_pred)

array([[31379,   589],
       [ 3276,   924]], dtype=int64)

# Testing Model Accuracy

In [27]:
# The model accuracy is calculated by (a+d)/(a+b+c+d)
(39156+1162)/(39156+766+4127+1162)

0.8917741257658535

In [29]:
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report

In [30]:
# for train data
print(classification_report(y_train,y_pred))

              precision    recall  f1-score   support

           0       0.91      0.98      0.94     31968
           1       0.61      0.22      0.32      4200

    accuracy                           0.89     36168
   macro avg       0.76      0.60      0.63     36168
weighted avg       0.87      0.89      0.87     36168



In [31]:
# for test data
print(classification_report(y_test,y_pred_test))

              precision    recall  f1-score   support

           0       0.90      0.98      0.94      7954
           1       0.58      0.22      0.32      1089

    accuracy                           0.89      9043
   macro avg       0.74      0.60      0.63      9043
weighted avg       0.86      0.89      0.86      9043



# The model is having 89% accuracy for the train data and having 89.3% accuracy for the Test data (data is not seen by the model). So the model is stable.